### PR TITLE
[Mime] switching source of mime types from Apache to "MIME-db"

### LINKED ISSUE #37412: Consider using another source for MIME types
**Description**  
PHP is switching from Apache to "MIME-db" because the Apache MIME type list is incomplete and it doesn't update frequently. See https://github.com/php/php-src/pull/5764

The differences are not huge, but for example, `.ini`, `.json5`, `.jsx`, `.less`, `.m4p` (<- not a typo) and other extensions are missing in our list but exist in MIME-db; others have different default types (e.g. `.gz`, `.xlf`, etc.)

### PR DESCRIPTION
**Pull request by desarrolla2**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #37412
| License       | MIT
| Doc PR        | no

### CODE REVIEW

**Comment by desarrolla2** [comment]
I preferred to use the "v1.44.0" tag instead of master because they recommend it, since there may be json format changes in the future.

**Comment by fabpot** [comment]
Great! Can you apply fabbot's patch? Also, can you run the script to see the diff with what we have now?

File: src/Symfony/Component/Mime/MimeTypes.php [hunk 1]
```diff
@@ -1110,11 +1185,11 @@ public function guessMimeType(string $path): ?string
         'audio/m3u' => ['m3u', 'm3u8', 'vlc'],
         'audio/m4a' => ['m4a', 'f4a'],
         'audio/midi' => ['mid', 'midi', 'kar', 'rmi'],
-        'audio/mobile-xmf' => ['xmf'],
+        'audio/mobile-xmf' => ['mxmf', 'xmf'],
         'audio/mp2' => ['mp2'],
         'audio/mp3' => ['mp3', 'mpga'],
         'audio/mp4' => ['m4a', 'mp4a', 'f4a'],
-        'audio/mpeg' => ['mp3', 'mpga', 'mp2', 'mp2a', 'm2a', 'm3a'],
+        'audio/mpeg' => ['mpga', 'mp2', 'mp2a', 'mp3', 'm2a', 'm3a'],
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Mime/MimeTypes.php:1192]: related to https://github.com/symfony/symfony/pull/37389

File: src/Symfony/Component/Mime/MimeTypes.php [hunk 2]
```diff
@@ -1460,7 +1587,7 @@ public function guessMimeType(string $path): ?string
         'text/x-lua' => ['lua'],
         'text/x-lyx' => ['lyx'],
         'text/x-makefile' => ['mk', 'mak'],
-        'text/x-markdown' => ['md', 'mkd', 'markdown'],
+        'text/x-markdown' => ['mkd', 'md', 'markdown'],
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Mime/MimeTypes.php:1590]: if the 1st is the most used, then `md` should be first?
does the order mean something in the source db?

File: src/Symfony/Component/Mime/MimeTypes.php [hunk 3]
```diff
@@ -334,18 +368,23 @@ public function guessMimeType(string $path): ?string
         'application/vnd.antix.game-component' => ['atx'],
         'application/vnd.appimage' => ['appimage'],
         'application/vnd.apple.installer+xml' => ['mpkg'],
-        'application/vnd.apple.keynote' => ['key'],
+        'application/vnd.apple.keynote' => ['keynote', 'key'],
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Mime/MimeTypes.php:371]: `key` should be first ?

**Comment by nicolas-grekas** [comment]
Thanks for working on this.
About my comments above related to ordering, maybe we should ask them on the upstream db?

File: src/Symfony/Component/Mime/MimeTypes.php (hunk 1 above)
Comment (YaFou) [reply src/Symfony/Component/Mime/MimeTypes.php:1192]: Yes, it's the most common format. You can also add `mp1` like defined [here (page3)](https://tools.ietf.org/html/rfc3003).

File: src/Symfony/Component/Mime/MimeTypes.php (hunk 2 above)
Comment (javiereguiluz) [reply src/Symfony/Component/Mime/MimeTypes.php:1590]: For reference, the source DB uses this preference for merging the multiple sources:

```
// source preference (least -> most)
var preference = ['nginx', 'apache', undefined, 'iana']
```

File: src/Symfony/Component/Mime/MimeTypes.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Mime/MimeTypes.php:1192]: @YaFou could you please report this to https://github.com/jshttp/mime-db? Maybe send them a PR also?

Comment (YaFou) [reply src/Symfony/Component/Mime/MimeTypes.php:1192]: I will do that this evening!

Comment (YaFou) [reply src/Symfony/Component/Mime/MimeTypes.php:1192]: @nicolas-grekas: see jshttp/mime-db#201.

File: src/Symfony/Component/Mime/Resources/bin/update_mime_types.php [hunk 4]
```diff
@@ -10,15 +10,13 @@
  */
 
 // load new map
-$data = file_get_contents('https://svn.apache.org/repos/asf/httpd/httpd/trunk/docs/conf/mime.types');
+$data = json_decode(file_get_contents('https://raw.githubusercontent.com/jshttp/mime-db/v1.44.0/db.json'), true);
```
Comment (YaFou) [inline src/Symfony/Component/Mime/Resources/bin/update_mime_types.php:13]: Wouldn't it be better to use the link provided in [the repository readme](https://github.com/jshttp/mime-db#database-download)?
```suggestion
$data = json_decode(file_get_contents('https://cdn.jsdelivr.net/gh/jshttp/mime-db@v1.44.0/db.json'), true);
```

Comment (desarrolla2) [reply src/Symfony/Component/Mime/Resources/bin/update_mime_types.php:13]: I'm not sure. Do you think that jsdelivr will have more availability that githubusercontent?

Comment (YaFou) [reply src/Symfony/Component/Mime/Resources/bin/update_mime_types.php:13]: No, I mean that the repository recommends this link. In addition, I think GitHub has not a better availability than JSdelivr, it's just my opinion *(and I don't have some totals)*.

File: src/Symfony/Component/Mime/Resources/bin/update_mime_types.php [hunk 5]
```diff
@@ -10,15 +10,13 @@
  */
 
 // load new map
-$data = file_get_contents('https://svn.apache.org/repos/asf/httpd/httpd/trunk/docs/conf/mime.types');
+$data = json_decode(file_get_contents('https://cdn.jsdelivr.net/gh/jshttp/mime-db@v1.44.0/db.json'), true);
 $new = [];
-foreach (explode("\n", $data) as $line) {
-    if (!$line || '#' == $line[0]) {
+foreach ($data as $mimeType => $mimeTypeInformation) {
+    if (!array_key_exists('extensions', $mimeTypeInformation)) {
```
Comment (juanmrad) [inline src/Symfony/Component/Mime/Resources/bin/update_mime_types.php:16]: `array_key_exists` is deprecated on PHP 7.4 shouldn't we use `isset` instead as it is available on all versions.?

Comment (nicolas-grekas) [reply src/Symfony/Component/Mime/Resources/bin/update_mime_types.php:16]: AFAIK, `array_key_exists` is not deprecated in 7.4 so it should be fine.

File: src/Symfony/Component/Mime/MimeTypes.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Mime/MimeTypes.php:1192]: @YaFou can you revert this change, maybe adding a comment like `// mp3 should be first`
same in other places where I commented about ordering?
that'd allow unlocking this PR

Comment (nicolas-grekas) [reply src/Symfony/Component/Mime/MimeTypes.php:1192]: I meant @desarrolla2 sorry :

File: src/Symfony/Component/Mime/Resources/bin/update_mime_types.php (hunk 5 above)
Comment (desarrolla2) [reply src/Symfony/Component/Mime/Resources/bin/update_mime_types.php:16]: I was looking for more info and it is deprecated to be used on objects, not for arrays. https://www.php.net/manual/en/migration74.deprecated.php#migration74.deprecated.core.array-key-exists-objects

File: src/Symfony/Component/Mime/MimeTypes.php (hunk 1 above)
Comment (desarrolla2) [reply src/Symfony/Component/Mime/MimeTypes.php:1192]: @nicolas-grekas do you want that:

a) I put the extensions in the order that you comment at MimeTypes.php
b) I update the script update_mime_types.php to force that order after download sources
c) I add comments with your suggestions at MimeTypes.php

Comment (nicolas-grekas) [reply src/Symfony/Component/Mime/MimeTypes.php:1192]: I think b) would be good. Then the comments would be found in update_mime_types.php only.

**Comment by desarrolla2** [comment]
should i rebase the commit to improve the history?

**Comment by stof** [comment]
@desarrolla2 yes please

File: src/Symfony/Component/Mime/MimeTypes.php [hunk 6]
```diff
@@ -2377,20 +2568,25 @@ public function guessMimeType(string $path): ?string
         'msf' => ['application/vnd.epson.msf'],
         'msg' => ['application/vnd.ms-outlook'],
         'msh' => ['model/mesh'],
-        'msi' => ['application/x-msdownload', 'application/x-msi'],
+        'msi' => ['application/octet-stream', 'application/x-msdownload', 'application/x-msi'],
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Mime/MimeTypes.php:2571]: I've checked the source and I think we should always list `'application/octet-stream'` last.
Alternatively, we should just skip it. The reason is that we already fallback to `'application/octet-stream'` by default in `DataPart`. For suren listing it first is going to be a regression.

Comment (desarrolla2) [reply src/Symfony/Component/Mime/MimeTypes.php:2571]: @nicolas-grekas you prefer that:

a) I put 'application/octet-stream' last.
b) I skip 'application/octet-stream'

Comment (nicolas-grekas) [reply src/Symfony/Component/Mime/MimeTypes.php:2571]: Skipping would be best IMHO

File: src/Symfony/Component/Mime/Resources/bin/update_mime_types.php [hunk 7]
```diff
@@ -66,6 +64,27 @@
 $map = array_replace_recursive($current, $new);
 ksort($map);
 
+// remove some extensions from the map
+$removeExtensionByMimeType = [
+    'application/octet-stream' => 'msi',
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Mime/Resources/bin/update_mime_types.php:69]: sorry I missed telling that my note about this should apply to all extensions, not only msi

**Comment by fabpot** [comment]
@desarrolla2 I've fixed your last commit.

**Review by fabpot — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @desarrolla2.

**Pull request merged**